PAYARA-3835 Cluster wide data aggregation for Monitoring Console

jbee opened the pull request:
Main changes in this PR:

* adds a module for the core of MC so that it can run within each instance
  * collects the data and does data exchange with the DAS
  * contains core data structures to represent data in memory (including some statistics)
* adds MC webapp module (is manually deployed)
* updates to metric collection due to changes in MP Metrics 2.0

**Dataset Representation**
A bit on the `SeriesDataset` implementation: It uses effectively immutable data to allow consistent snapshots across all series. This mean when adding a point - time + value - to a dataset a new instance is returned representing the change. The returned implementation depends on how the point fits into the points observed so far. If the value is unchanged the `ConstantDataset` is used. If it changes the `PartialDataset` is used. If it seems to become stable again the `StableDataset` is used.

So depending on the value of the series over time we get: 
```
EmptyDataset => ConstantDataset
EmptyDataset => ConstantDataset => PartialDataset
EmptyDataset => ConstantDataset => PartialDataset => StableDataset
EmptyDataset => ConstantDataset => PartialDataset => StableDataset => PartialDataset ...
```

**Data Exchange**
The exchange happens in `InMemoryMonitoringDataRepository` that uses a hazelcast topic to exchange the data. A message `SeriesDatasetsSnapshot` contains all values of the collected series for the collected time. When the message arrives at the listening DAS it is added to the in memory store. Only the DAS stores data. Since members are aware of listening members for a topic the message is only send to the DAS in case it is available. This keeps network traffic as minimal as it can be. The overall network traffic does not increase when length of the in-memory store is enlarged. It increases linear with the number of metric series and instances in the cluster.

jbee commented:
jenkins test please

jbee commented:
@Cousjava that is very unfortunate. Must be related to OSGi or something since I did not change anything AMX related directly in this PR. Have to look into it.

Pandrex247 commented on appserver/monitoring-console/webapp/src/main/java/fish/payara/monitoring/web/SeriesStatistics.java:
```diff
@@ -0,0 +1,44 @@
+package fish.payara.monitoring.web;
```
Copyright

Pandrex247 commented on appserver/monitoring-console/webapp/src/main/webapp/index.html:
```diff
@@ -0,0 +1,216 @@
+<!DOCTYPE html>
```
Copyright...?

Pandrex247 commented:
Jenkins seemed to hang trying to complete the quicklook tests - I'll run it again

Pandrex247 commented:
Jenkins test please

Pandrex247 commented:
I also get the NPE:
* Start domain and load Admin Console
* Enable AMX and change monitoring level of all settings to HIGH
* Save config and wait for AMX to try and start (no server restart necessary)

Pandrex247 commented:
It appears to start fine after a server restart, presumably it only throws the NPE when dynamically enabling AMX without a restart

jbee commented on appserver/featuresets/payara-web/pom.xml:
```diff
@@ -306,5 +306,13 @@
+        <!-- Monitoring Console package -->
+        <dependency>
+            <groupId>org.glassfish.main.packager</groupId>
+            <artifactId>monitoring-console</artifactId>
```
I'd also think that for now less impact is better. I move it to server.

jbee commented:
The problem is that `DomainRoot` proxy is `null` - starting to think that the potential for this has always been there but due to some addition this is not triggered reliably. The bean causing this is `amx:pp=,type=domain-root` itself. Maybe the code should not even run for this particular bean?

jbee commented:
jenkins test please

jGauravGupta commented:
Tested with latest commits/merge but NPE is still reproducible.
```
  AMX Startup Service: Error loading AMX Beans.
java.lang.RuntimeException: java.lang.NullPointerException
	at org.glassfish.admin.amx.impl.AMXStartupService.loadDomainRoot(AMXStartupService.java:251)
	at org.glassfish.admin.amx.impl.AMXStartupService._loadAMXMBeans(AMXStartupService.java:327)
	at org.glassfish.admin.amx.impl.AMXStartupService.loadAMXMBeans(AMXStartupService.java:224)
.............
	at java.management/com.sun.jmx.mbeanserver.JmxMBeanServer.invoke(JmxMBeanServer.java:801)
	at com.sun.enterprise.v3.admin.DynamicInterceptor.invoke(DynamicInterceptor.java:310)
	at org.glassfish.admin.mbeanserver.BootAMX.bootAMX(BootAMX.java:132)
	at fish.payara.admin.amx.AMXBootService.startup(AMXBootService.java:87)
	at fish.payara.admin.amx.AMXBootService.setEnabled(AMXBootService.java:93)
	at fish.payara.admin.amx.AMXBootService.changed(AMXBootService.java:106)
	at org.jvnet.hk2.config.Transactions$ConfigListenerJob.process(Transactions.java:402)
```

jGauravGupta commented on appserver/packager/monitoring-console/src/main/assembly/monitoring-console.xml:
```diff
@@ -0,0 +1,64 @@
+<!--
+  DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
+
+  Copyright (c) [2017-2018] Payara Foundation and/or its affiliates. 
```
Wrong copyright  year.

jGauravGupta commented on nucleus/payara-modules/hazelcast-bootstrap/src/main/java/fish/payara/nucleus/store/ClusteredStore.java:
```diff
@@ -38,13 +38,13 @@
  */
```
Copyright year should be updated.

jGauravGupta commented on appserver/monitoring-console/webapp/src/main/java/fish/payara/monitoring/web/MonitoringConsoleResouce.java:
```diff
@@ -0,0 +1,93 @@
+@Path("/")
+@Produces(MediaType.APPLICATION_JSON)
+@RequestScoped
+public class MonitoringConsoleResouce {
```
Typo in the class name `MonitoringConsoleResouce `

jbee commented:
jenkins test please

jbee commented:
@Pandrex247 , @jGauravGupta , @Cousjava I addressed PR comments. The NPE is some sort of initialisation order issue that is triggered by AMX being also a monitoring data source. Since I did not find a solution I opted to remove AMX as source for now. This is a workaround but AMX as source is not essential and will allow us to test monitoring console in general. AMX source needs to be added in a later release.

Pandrex247 commented:
The issue with quicklook seems to have fixed itself
